File: DM&ADS/formal languages/regex language automatons/F.java
package DiskrLab5;

import java.io.*;
import java.util.*;

public class F {
    boolean result = true;
    boolean[] visited;
    HashSet<Integer> terminals1 = new HashSet<>();
    HashMap<Integer, HashMap<Character, Integer>> map = new HashMap<>();
    HashSet<Integer> terminals2 = new HashSet<>();
    HashMap<Integer, HashMap<Character, Integer>> map2 = new HashMap<>();


    void dfs(int v1, int v2) {
        visited[v1] = true;
        if (terminals1.contains(v1) != terminals2.contains(v2)) {
            result = false;
            return;
        }

        HashMap<Character, Integer> curMap = map.get(v1);
        if (curMap == null && map2.get(v2) != null) {
            result = false;
            return;
        }
        if (curMap != null) {
            for (Character ch : curMap.keySet()) {
                int u1 = map.get(v1).get(ch);
                if (!map2.get(v2).containsKey(ch)) {
                    result = false;
                    return;
                }
                int u2 = map2.get(v2).get(ch);
                if((u1==v1 && u2!=v2)||(v1!=u1 && u2==v2)){
                    result=false;
                    return;
                }
                if (!visited[u1]) {
                    dfs(u1, u2);
                }
            }
        }
        return;
    }

    void solve(PrintWriter out) throws IOException {
        FastScanner in = new FastScanner("isomorphism.in");
        int n1 = in.nextInt();
        int m1 = in.nextInt();
        int k1 = in.nextInt();

        for (int i = 0; i < k1; i++) {
            terminals1.add(in.nextInt());
        }

        for (int i = 0; i < m1; i++) {
            int from = in.nextInt();
            int to = in.nextInt();
            String a = in.next();
            if (!map.containsKey(from)) {
                map.put(from, new HashMap<>());
                map.get(from).put(a.charAt(0), to);

            } else {
                map.get(from).put(a.charAt(0), to);
            }
        }

        int n2 = in.nextInt();
        int m2 = in.nextInt();
        int k2 = in.nextInt();

        for (int i = 0; i < k2; i++) {
            terminals2.add(in.nextInt());
        }

        for (int i = 0; i < m2; i++) {
            int from = in.nextInt();
            int to = in.nextInt();
            String a = in.next();
            if (!map2.containsKey(from)) {
                map2.put(from, new HashMap<>());
                map2.get(from).put(a.charAt(0), to);

            } else {
                map2.get(from).put(a.charAt(0), to);
            }
        }

        if (n1 != n2) {
            result = false;
        }
        if (m1 != m2)
            result = false;
        if (k1 != k2)
            result = false;
        if(map.size()!=map2.size()){
            result=false;
        }
        if (!result) {
            out.println("NO");
            return;
        }
        visited = new boolean[n1 + 1];
        dfs(1, 1);
        if (result) out.println("YES");
        else out.println("NO");


    }


    StreamTokenizer in;
    PrintWriter out;

    void run() throws IOException {
        out = new PrintWriter("isomorphism.out");
        solve(out);
        out.flush();
    }

    public static void main(String[] args) throws IOException {
        new F().run();

    }

    class FastScanner {
        BufferedReader reader;
        StringTokenizer tokenizer;

        public FastScanner(String fileName) throws IOException {
            reader = new BufferedReader(new FileReader(fileName));
        }

        public FastScanner() throws IOException {
            reader = new BufferedReader(new InputStreamReader(System.in));
        }

        public String next() throws IOException {
            while (tokenizer == null || !tokenizer.hasMoreTokens()) {
                String line = reader.readLine();
                if (line == null) {
                    return null;
                }
                if (line.equals("")) {
                    return "";
                }
                tokenizer = new StringTokenizer(line);
            }
            return tokenizer.nextToken();
        }

        public int nextInt() throws IOException {
            return Integer.parseInt(next());
        }

        public double nextDouble() throws IOException {
            return Double.parseDouble(next());
        }
    }

}

File: DM&ADS/genfuctions/F.kt
package diskrLab10

import java.io.*
import java.util.*


object F {
    const val fileName = ""

    //================================================================================
    val MOD: Long = 104857601
    var k = 0
    var n: Long = 0
    lateinit var start: LongArray
    lateinit var coeff: LongArray
    fun run() {
        k = IO.nextInt()
        n = IO.nextLong() - 1
        start = LongArray(2 * k)
        coeff = LongArray(k)
        var q = LongArray(k + 1)
        for (i in 0 until k) {
            start[i] = IO.nextLong()
        }
        q[0] = 1
        for (i in 0 until k) {
            val c = IO.nextLong()
            coeff[i] = c
            q[i + 1] = (-c % MOD + MOD) % MOD
        }
        while (n >= k) {
            for (i in k until 2 * k) {
                for (j in 1..k) {
                    start[i] += ((-q[j] % MOD + MOD) % MOD * start[i - j]) % MOD
                    start[i] %= MOD
                }
            }
            // q2 = q(-t)
            val q2 = q.copyOf()
            for (i in 1..k step 2) {
                q2[i] = (-q2[i] % MOD + MOD) % MOD
            }
            val mult = LongArray(2 * (k + 1))
            for (i in 0..k) {
                for (j in 0..k) {
                    mult[i + j] += (q[i] * q2[j]) % MOD
                    mult[i + j] = (mult[i + j] % MOD + MOD) % MOD
                }
            }
            start = start.filterIndexed { ind, _ -> ind.toLong() % 2 == n % 2 }.toLongArray().copyOf(2 * k)
            q = mult.filter { it != 0L }.toLongArray()
            n /= 2
        }
        IO.println(start[n.toInt()])

    }
    //==================================================================================

    @JvmStatic
    fun main(args: Array<String>) {
        this.run()
        IO.writer.flush()
        IO.reader.close()

    }

    object IO {
        val reader: BufferedReader
        val writer: PrintWriter
        private var tokenizer: StringTokenizer? = null

        init {
            if (fileName.isEmpty()) {
                reader = BufferedReader(InputStreamReader(System.`in`))
                writer = PrintWriter(System.out)
            } else {
                reader = BufferedReader(FileReader("$fileName.in"))
                writer = PrintWriter(FileWriter("$fileName.out"))
            }
        }

        fun next(): String {
            while (tokenizer == null || !tokenizer!!.hasMoreTokens()) {
                val line = reader.readLine() ?: throw NullPointerException()
                if (line == "") {
                    return ""
                }
                tokenizer = StringTokenizer(line)
            }
            return tokenizer!!.nextToken()
        }

        fun nextLine(): String {
            return reader.readLine()
        }

        fun nextInt(): Int {
            return next().toInt()
        }

        fun nextDouble(): Double {
            return next().toDouble()
        }

        fun nextLong(): Long {
            return next().toLong()
        }

        fun <T> print(obj: T) {
            writer.print(obj)
        }

        fun <T> println(obj: T) {
            writer.println(obj)
        }
    }
}